App/HostedReactor.fs

```diff
@@ -0,0 +1,67 @@
+module HostedReactor
+
+open System
+open Microsoft.Extensions.Hosting
+open Propulsion.Internal
+open Propulsion.MessageDb
+
+type HostedReactor(connectionString, log, service: InvoiceNumberingReactor.Service, checkpointService: CheckpointStore.CheckpointService) =
+  inherit BackgroundService()
+
+  let stats =
+    { new Propulsion.Streams.Stats<_>(log, TimeSpan.FromMinutes 1, TimeSpan.FromMinutes 1) with
+        member _.HandleOk x = ()
+        member _.HandleExn(log, x) = () }
+
+  [<Literal>]
+  let groupName = "InvoiceNumberingReactor"
+
+  let categories = [| Invoice.Category |]
+
+  let checkpoints = CheckpointStore.CheckpointStore(checkpointService, groupName, TimeSpan.FromSeconds 10)
+  let maxReadAhead = 100
+  let maxConcurrentStreams = 2
+
+  let handle struct(stream, events) =
+    async {
+      for event in events do
+        do! service.Handle(stream, event)
+
+      return struct (Propulsion.Streams.SpanResult.AllProcessed, ())
```

Comment:
could add stats about 1) numbers genned 2) invoices numbered 3) replays (neither)
(the high level stats will show the total requests, streams and events traversed, so in the case of the event not being relevant, you can return an 'Outcome.Noop', but no need to gather stats on it)

Reply:
The stats area is one I'm not very familiar/comfortable with. Would love to learn more about how you approach these things!

Reply:
TL;DR the default wiring is plumbed for each Handler invocation (which returns `SpanResult*'Outcome`) to yield an `'Outcome` which summarises the work done for one invocation
That outcome is flowed through to the `HandleOk` on the `Stats` after successful processing (the calls to it are serialized so no concurrency control required)
The stats are then dumped every 1m (again with a guarantee of no concurrent HandleOk/HandleExn calls at the same time)
Many of the dotnet-templates have examples, and/or one can look at the Propulsion internal ones for slightly more exotic schemes.
A struct DU is typically a good fit for said Outcome
A secondary function of this is to enable a clear way for Hander outcomes to be emitted to e.g. Prometheus - at the same time as one is gathering stats for local use, you also map that to a Prometheus metric. Example:

- https://github.com/jet/dotnet-templates/blob/4ab931c605b04e5ad35f4208c6a82085fa302719/propulsion-cosmos-reactor/ReactorMetrics.fs#L1
- https://github.com/jet/dotnet-templates/blob/4ab931c605b04e5ad35f4208c6a82085fa302719/propulsion-cosmos-reactor/Reactor.fs#L12